### ISSUE TITLE
PRD: Proxy Support on JetBrains

### ISSUE THREAD

**Issue by aramaraju** [sourcegraph/jetbrains#1127]
### Description:

Many modern enterprises deploy a proxy between their developers' devices and the internet, allowing them to deny access to unauthorized websites, protect their exact IP addresses of their workforce and inspect the traffic for any cyberthreat (malware, phishing attacks, etc). Today, any Cody specific traffic from JetBrains cannot be successfully directed to the relevant endpoints through a proxy, thereby affecting the Cody purchase for such enterprises. 

### Objectives:

- Ability to use the Cody plugin on JetBrains in an enterprise setting, where all network traffic is routed through a proxy
- Improve the visibility and logging to make it easier to diagnose and fix any issues

### Acceptance Criteria:

1. Ability to use Cody with JetBrains for any supported OS platforms (Windows, Mac and Linux) under the following operating modes:
- Connect directly without a proxy
- Connect to a proxy by automatically detecting system proxy settings with proxy environmental variables ([relevant instructions](https://docs.cloudfoundry.org/cf-cli/http-proxy.html))
- Connect to a proxy  by automatically detecting system proxy settings specified at the OS  settings
![CleanShot 2024-03-18 at 12 52 21@2x](https://github.com/sourcegraph/jetbrains/assets/51247042/2f318f13-8db8-462b-a811-981690b398ac)
- Connect to a proxy by automatically detecting settings from a proxy auto-config (PAC) file, where the (URL) location of the PAC file is specified 
![CleanShot 2024-03-18 at 13 00 06@2x](https://github.com/sourcegraph/jetbrains/assets/51247042/5fcd32ea-01e4-4081-ad9c-6b8b78d759e7)
- Connect to a proxy with manual settings (using Host name, port number, exclusions) on the IDE settings pane ([Path](jetbrains://idea/settings?name=Appearance+%26+Behavior--System+Settings--HTTP+Proxy) 
![CleanShot 2024-03-18 at 12 56 26@2x](https://github.com/sourcegraph/jetbrains/assets/51247042/d15f55ae-7ddf-4021-b3fa-678136e7d5f5)
2. The following types of proxies should be supported: HTTP, HTTPS, SOCKS
3. If there are no proxy configuration specified in the IDE settings, the environmental variables/OS settings must be checked first and honored for all Cody traffic by default
4. If there are conflicts between the system level settings and the IDE settings, the IDE settings should take precedence
5. The user and device must be able to authenticate with the proxy server in one of the following ways:
- Using credentials provided through the environment path variables (example: http://user:password@proxyIP:proxyPort/)
- Using credentials provided through the IDE settings 
![CleanShot 2024-03-18 at 12 51 16@2x](https://github.com/sourcegraph/jetbrains/assets/51247042/65cccb9b-0f43-4923-8cf0-b39bd48aa260)
6. Cody must be able to read any custom certificates already installed on the user's device ([related tracking issue](https://github.com/sourcegraph/sourcegraph/issues/61021)  
7. All network traffic related to Cody or Sourcegraph must be sent through the proxy when a proxy is configured. This includes, but is not limited to Cody traffic to LLM providers telemetry and any additional binaries required for Cody plugin (today we download them from github.com directly, refer to this [RFC](https://docs.google.com/document/d/1AsnoSOoLjRq-iobfzl4d1ckOg36kL3yZZtvo3UriwBE/edit#heading=h.trqab8y0kufp) for additional discussion)
8. Provide an ability for the admin to easily ping an endpoint to quickly see if they have setup their proxy configuration correctly. They should also be able to do so using the 'Check Connection' setting here:
![CleanShot 2024-03-18 at 13 23 45@2x](https://github.com/sourcegraph/jetbrains/assets/51247042/ca199389-8c7b-45d2-a090-e41fe98c4ccf)
9. When Sourcegraph/Cody relevant traffic cannot be reached, an appropriate number of retry attempts must be done with an eventual timeout. The user must be able to continue using the IDE for other non-Cody related tasks during the retries and after timeout
10. When there is an issue with Cody being able to reach the relevant endpoints, there must be appropriate error messages suggesting a network reachability error. If there is a proxy setup, this information must be indicated in these error messages.
11.  If possible, the IntelliJ logs must at least identify the following:
- Whether the client is able to connect with the appropriate endpoints
- What exact traffic was blocked/couldn't be reached
- If there is a failure, provide additional details into whether the failure was due to an authentication or certificate or reachability/availability error (where the endpoint itself is down)
12. If the admin/enterprise prefers to forgo any inspection through the proxy (or firewall), a complete list of URLs/domains must be available for them to whitelist any Cody specific traffic. 
14. Relevant documentation must capture the aforementioned information for deploying and troubleshooting proxy settings

## Open questions
1. Does the above specified cert fix work for transparent proxies? 

### Design Tasks:

```[tasklist]
### Design Tasks
- [ ] Add a draft title or issue reference here
```

### Engineering Tasks:

```[tasklist]
### Engineering Tasks
- [ ] https://github.com/sourcegraph/cody/pull/3695
```

**Comment by kalanchan** [comment]
Further scoping:
1. Testing criteria: simple local proxy is fine
1. Accepts OS level proxy settings (more important and needs more eng discovery)
   a. understand how this works in VS Code and why(?) it doesn't in Agent
1. Accepts IDE level proxy settings
   a. HTTP/HTTPS
   b. SOCKS
   c. Authentication

**Comment by kalanchan** [comment]
@pkukielka to do initial discovery, then we will sync with Enterprise to see how much work we have to do

**Comment by dominiccooney** [comment]
From the extension side, here's the proxy setup we use for `fetch`: [`initializeNetworkAgent`](https://sourcegraph.com/github.com/sourcegraph/cody@a0056f25730b06a250ba2b667f04111bb7666ae8/-/blob/vscode/src/fetch.node.ts?L47) but everything else looks like VSCode defaults.

VSCode's proxy value add over node. Agent would need to replicate the effects of this:
- Configuration options: [docs](https://sourcegraph.com/github.com/microsoft/vscode-docs@51978c202dc887e2e142ca6979e5825ffb3a164a/-/blob/docs/getstarted/settings.md?L2412), [code](https://sourcegraph.com/github.com/microsoft/vscode@e49b522e12d1a55dfe2950af355767c9d5b824aa/-/blob/src/vs/platform/request/common/request.ts?L131)
- [Extension host applying the config](https://sourcegraph.com/github.com/microsoft/vscode@a2acd131e47500cf4bd7d602626f0b54ab266904/-/blob/src/vs/workbench/api/node/extHostExtensionService.ts?L88:9-88:29), by [patching node to hook bits of http, https](https://sourcegraph.com/github.com/microsoft/vscode@a2acd131e47500cf4bd7d602626f0b54ab266904/-/blob/src/vs/workbench/api/node/proxyResolver.ts?L76)

TL;DR code: VSCode has extensive options for proxies, and configures proxies for the extension. That's probably why networking works in the VSCode extension but not Agent. @pkukielka 's investigation of the JetBrains direct GraphQL queries are still relevant.

**Comment by pkukielka** [comment]
I think that line might be actually the key and the entry point for us.
I suspect it evaluates to `null` in the agent, because both `_VSCODE_NODE_MODULES` and `@vscode/proxy-agent/out/agent` are not present in the agent scope. But I think if we add explicit dependency on the `@vscode/proxy-agent` it will be then visible and it will be picked up.
I also expect that `new https.Agent` in fact returns the instance of the `vscode-proxy-agent` due to the hacks in vscode linked by @dominiccooney. We might want to utilize `getCustomAgent` to create custom instance for us in case of agent.
IMHO there is a good chance that those two modifications will make it work (although we might still want to pass proxy settings from the IntelliJ settings).

**Comment by pkukielka** [comment]
As for investigation of the JetBrains direct GraphQL queries it turns out they are indeed using custom HTTP query mechanism. 
On the flip side, under the hood they are using `com.intellij.util.io.HttpRequests` which seemed to have build-in support for proxies defined in IntelliJ:

```java
  @RequiresBackgroundThread
  @Throws(IOException::class, ProcessCanceledException::class)
  fun <T> execute(indicator: ProgressIndicator, request: SourcegraphApiRequest<T>): T {
    indicator.checkCanceled()
    return createRequestBuilder(request)
        .tuner { connection ->
          request.additionalHeaders.forEach(connection::addRequestProperty)
          connection.addRequestProperty(HttpSecurityUtil.AUTHORIZATION_HEADER_NAME, "token $token")
        }
        .useProxy(useProxy)  // <--------------- [[[ LOOK HERE ]]]
        .execute(request, indicator)
  }
```

I was able to test it for HTTP protocol queries using `SquidMan` the following firewall configuration:

```
# 1. Install squidman: brew install --cask squidman
# 2. Setup squidman to work on 8080 HTTP port
# 3. Apply this firewall config: sudo pfctl -e -f proxy_firewall.conf
# 4. To disable it: sudo pfctl -d

# Set the interface to be used
if="en0"

# Default Deny Policy, change to ` { 80, 443 }` is you want to also block HTTPS
block out quick proto tcp from any to any port { 80 }

# Skip the loop back interface
set skip on lo

# Set http proxy (8080) port
web_ports = "{8080}"

```

And I can confirm **it works**. I mean if you define proxy parameters in the IntelliJ settings they are properly applied to our GraphQL connections.
I'm in progress of investigating [testing HTTPS](https://wiki.squid-cache.org/Features/HTTPS).

**Comment by arafatkatze** [comment]
There were testing issues with [add support for proxies PR ](https://github.com/sourcegraph/cody/pull/3769) so it had to be [reverted](https://github.com/sourcegraph/cody/pull/3814)  but I will issue a fix for that this week.

**Comment by arafatkatze** [comment]
Closing this issue now [see this comment ](https://github.com/sourcegraph/cody-issues/issues/72#issuecomment-2115156877) on a related issue

**Comment by arafatkatze** [comment]
NOTE: I still have to send testing instructions to Kalan and Aravind to test things but that is just a sanity check which will work so I am not too worried about it.

**Comment by kalanchan** [comment]
@arafatkatze just for completeness sakes, can you close the issue when the testing instructions and docs are ready and link them to this thread here? 

it'll be good to have all the information in this ticket history incase we need to refer back to it!

**Comment by arafatkatze** [comment]
[Proxy Testing Guide](docs.google.com/document/d/1j1seazcli9_nclys1rt24hnheepi9qdztunoutbvvdg/edit)

[Proxy Setup Docs](https://github.com/sourcegraph/docs/pull/342#issuecomment-2125428046)
